### PR TITLE
multi-room: add snapweb to snapcast server

### PR DESCRIPTION
**Pull request by ejannett**
This adds the simple web application snapweb to the snapcast server.
The application allows to set properties of snapcast (multiroom) client
such as volume

Change-type: minor
Signed-off-by: emmanuel.jannetti@gmail.com
Connects-to: #319

### CODE REVIEW

**Comment by ejannett** [comment]
Hi,
I see that ResinCI/balena/build failed but trying to get details lead me to nowhere. 
Any help would be appreciated.

thx in advance

**Comment by ejannett** [comment]
Hi,

Can. someone explain me how to fix the build failure . The "Details" link send me to https://ci.balena-dev.com/. 
but form there login Is impossible for me.

thank in advance

EJ

**Comment by tmigone** [comment]
Hi @ejannett, first of all, thanks for the contribution! Apologies for the slow response here, this past two weeks we were having our anual balena team summit, so that took a lot of my time 😟 

About `snapweb`, I remember testing it a few months back, the reason we haven't included it in balenaSound is because we want to build our own UI which will include more controls than what `snapweb` can offer (not only snapcast stuff but other things related to balenaSound). We haven't settled on what that looks like yet though and it's still a few months out. I think for the time being it's a good idea to include `snapweb` so at least we have something, so thanks for bringing it back to my attention 👍 Just now that it's very likely that we will replace it in the future whenever we get something more comprehensive out.

About the CI error, the logs are unfortunately private (that's how concourse works). From what I can tell it looks like our builders were having some sort of temporary problem. Rerunning the tests with ((at)balena-ci retest) should do the trick. I'll do it now.

**Comment by tmigone** [comment]
@balena-ci rebase

**Comment by tmigone** [comment]
@balena-ci retest

**Comment by tmigone** [comment]
All right, I figured it out. Our builders don't do git submodule resolution so it won't build the image properly. I think the best solution is to pull the repo using `git clone`. I'll leave a few comments on the code to clarify.

File: core/multiroom/server/Dockerfile.template
Comment (tmigone) [inline core/multiroom/server/Dockerfile.template:8]:
```suggestion
RUN git clone https://github.com/badaix/snapweb.git snapweb
```

File: core/multiroom/server/Dockerfile.template
Comment (tmigone) [inline core/multiroom/server/Dockerfile.template:5]:
```suggestion
RUN install_packages snapcast-server git
```

**Review by tmigone — CHANGES_REQUESTED** [review]
Proposed changes to fix the builder problem with git modules. The `.gitmodules` and `snapweb` link to repo should be removed too I think.

**Comment by ejannett** [comment]
Hi tomas,
Thank for the info . All you've said make . I'll make the change.
I would be very interested to participate to the UI sub-project if need extra hands.

regards
manu

File: core/multiroom/server/Dockerfile.template
Comment (ejannett) [reply core/multiroom/server/Dockerfile.template:5]: change applied

File: core/multiroom/server/Dockerfile.template
Comment (ejannett) [reply core/multiroom/server/Dockerfile.template:8]: change applied

**Comment by ejannett** [comment]
Hi @tmigone  it seems that build is failing again. Any chance to have access to the logs ?
thank you

**Comment by thepont** [comment]
@ejannett I don't know if this is what's happening on the build server, but I just attempted to pull this in ( Because I really would like this feature as well ) and I am seeing the following error while attempting to push to balena cloud

```
multiroom-server]  Cloning into 'snapweb'...
[multiroom-server]  
[multiroom-server]  Removing intermediate container 5c506d5dc24b
[multiroom-server]   ---> 66200f3e14de
[multiroom-server]  Step 7/10 : RUN npm install --global --no-save typescript
[multiroom-server]   ---> Running in 17886cf52773
[multiroom-server]  /usr/bin/tsc -> /usr/lib/node_modules/typescript/bin/tsc
[multiroom-server]  /usr/bin/tsserver -> /usr/lib/node_modules/typescript/bin/tsserver
[multiroom-server]  + typescript@4.0.5
[multiroom-server]  added 1 package from 1 contributor in 3.103s
[multiroom-server]  Removing intermediate container 17886cf52773
[multiroom-server]   ---> bcf56d105be1
[multiroom-server]  Step 8/10 : RUN cd snapweb && make
[multiroom-server]   ---> Running in 3eb1ea98581c
[multiroom-server]  tsc --build page/tsconfig.json
[multiroom-server]  rm -rf dist
[multiroom-server]  mkdir dist
[multiroom-server]  mkdir dist/3rd-party
[multiroom-server]  cp -r page/3rd-party/libflac.js dist/3rd-party/
[multiroom-server]  cp page/favicon.ico dist
[multiroom-server]  cp page/index.html dist
[multiroom-server]  cp page/launcher-icon.png dist
[multiroom-server]  cp page/manifest.json dist
[multiroom-server]  cp page/mute_icon.png dist
[multiroom-server]  cp page/play.png dist
[multiroom-server]  mv page/snapcontrol.js dist
[multiroom-server]  mv page/snapcontrol.js.map dist
[multiroom-server]  mv page/snapstream.js dist
[multiroom-server]  mv page/snapstream.js.map dist
[multiroom-server]  cp page/speaker_icon.png dist
[multiroom-server]  cp page/stop.png dist
[multiroom-server]  cp page/styles.css dist
[multiroom-server]  Removing intermediate container 3eb1ea98581c
[multiroom-server]   ---> 3006337b5bf7
[multiroom-server]  Step 9/10 : COPY snapweb/dist /var/www
[multiroom-server]  COPY failed: stat /var/lib/balenav6l/tmp/docker-builder111250351/snapweb/dist: no such file or directory
[Info]              Uploading images
[Success]           Successfully uploaded images
[Error]             Some services failed to build:
[Error]               Service: multiroom-server
[Error]                 Error: COPY failed: stat /var/lib/balenav6l/tmp/docker-builder111250351/snapweb/dist: no such file or directory
[Info]              Built on arm03
[Error]             Not deploying release.

remote: error: hook declined to update refs/heads/master
To git.balena-cloud.com:gh_thepont/balenasound.git
 ! [remote rejected] master -> master (hook declined)
error: failed to push some refs to 'git.balena-cloud.com:gh_thepont/balenasound.git'
```

File: core/multiroom/server/Dockerfile.template
Comment (thepont) [inline core/multiroom/server/Dockerfile.template:10]: I think this should be 

`RUN cd snapweb && make && cd ..`

**Comment by tmigone** [comment]
@thepont is right! The build check on our CI runs `balena push <app>` and tries to build the app from there. Whenever that fails you can easily replicate it by trying a push from a freshly cloned repo.

Replacing `COPY snapweb/dist /var/www` with `RUN cd snapweb && make && mkdir -p /var/www && mv dist/* /var/www` should work.

**Comment by tmigone** [comment]
> Hi tomas,
> Thank for the info . All you've said regarding th UI make sense. About my PR, I'll make the changes.
> I would be very interested to participate to the UI sub-project if need extra hands. Does it already exists ?
> 
> regards
> manu

As for the UI piece, we haven't started that yet. I've just created #336, you can follow notifications over there so you are aware when we start that. And yeah, all hands are welcome!

**Comment by ejannett** [comment]
> @ejannett I don't know if this is what's happening on the build server, but I just attempted to pull this in into my local branch (Because I really would like this feature as well) and I am seeing the following error while attempting to push to balena cloud
> 
> ```
> multiroom-server]  Cloning into 'snapweb'...
> [multiroom-server]  
> [multiroom-server]  Removing intermediate container 5c506d5dc24b
> [multiroom-server]   ---> 66200f3e14de
> [multiroom-server]  Step 7/10 : RUN npm install --global --no-save typescript
> [multiroom-server]   ---> Running in 17886cf52773
> [multiroom-server]  /usr/bin/tsc -> /usr/lib/node_modules/typescript/bin/tsc
> [multiroom-server]  /usr/bin/tsserver -> /usr/lib/node_modules/typescript/bin/tsserver
> [multiroom-server]  + typescript@4.0.5
> [multiroom-server]  added 1 package from 1 contributor in 3.103s
> [multiroom-server]  Removing intermediate container 17886cf52773
> [multiroom-server]   ---> bcf56d105be1
> [multiroom-server]  Step 8/10 : RUN cd snapweb && make
> [multiroom-server]   ---> Running in 3eb1ea98581c
> [multiroom-server]  tsc --build page/tsconfig.json
> [multiroom-server]  rm -rf dist
> [multiroom-server]  mkdir dist
> [multiroom-server]  mkdir dist/3rd-party
> [multiroom-server]  cp -r page/3rd-party/libflac.js dist/3rd-party/
> [multiroom-server]  cp page/favicon.ico dist
> [multiroom-server]  cp page/index.html dist
> [multiroom-server]  cp page/launcher-icon.png dist
> [multiroom-server]  cp page/manifest.json dist
> [multiroom-server]  cp page/mute_icon.png dist
> [multiroom-server]  cp page/play.png dist
> [multiroom-server]  mv page/snapcontrol.js dist
> [multiroom-server]  mv page/snapcontrol.js.map dist
> [multiroom-server]  mv page/snapstream.js dist
> [multiroom-server]  mv page/snapstream.js.map dist
> [multiroom-server]  cp page/speaker_icon.png dist
> [multiroom-server]  cp page/stop.png dist
> [multiroom-server]  cp page/styles.css dist
> [multiroom-server]  Removing intermediate container 3eb1ea98581c
> [multiroom-server]   ---> 3006337b5bf7
> [multiroom-server]  Step 9/10 : COPY snapweb/dist /var/www
> [multiroom-server]  COPY failed: stat /var/lib/balenav6l/tmp/docker-builder111250351/snapweb/dist: no such file or directory
> [Info]              Uploading images
> [Success]           Successfully uploaded images
> [Error]             Some services failed to build:
> [Error]               Service: multiroom-server
> [Error]                 Error: COPY failed: stat /var/lib/balenav6l/tmp/docker-builder111250351/snapweb/dist: no such file or directory
> [Info]              Built on arm03
> [Error]             Not deploying release.
> 
> remote: error: hook declined to update refs/heads/master
> To git.balena-cloud.com:gh_thepont/balenasound.git
>  ! [remote rejected] master -> master (hook declined)
> error: failed to push some refs to 'git.balena-cloud.com:gh_thepont/balenasound.git'
> ```
> 
> I think this might be happening for the build machine and me but not you is because you build snapweb on your local machine before in the corresponding directory. The `COPY` is attempting to copy the file from the repository (or local machine) to the docker container, and since it doesn't exists it's failing. The copy has to be done in the run command.
> 
> `RUN cd snapweb && make && mkdir -p /var/www && mv dist/* /var/www` works

thank !  . good catch .  Testing locally , I had the right command.

**Comment by tmigone** [comment]
@ejannett looks good, thanks!

Before merging we'll need you to rebase your branch, what we want is to get all your commits squashed into a single commit and applied on top of the latest changes from our master branch. Reasoning for this is we don't want to clutter the commit history with several small commits for a minor/patch feature and also have multiple `Merge branch 'master' into ...` messages throughout the history.

Remember the single commit that remains must follow the [commit guidelines](https://sound.balenalabs.io/contributing#commit-guidelines) described in our docs (your first commit already does!).

The process to do this is not clearly documented at the moment in our docs so I'll leave it here, let me know if you need assistance:

```
git checkout <your-branch>
git remote add upstream https://github.com/balenalabs/balena-sound.git
git fetch upstream
git rebase -i upstream/master
git push origin <your-branch> --force-with-lease
```

When rebasing, if you made multiple commits then you'll need to `pick` the first one and `squash` the rest. Be sure to clean the commit message and only leave the relevant text.

This is how it should **not** look like:
```
# This is a combination of 3 commits.
# This is the 1st commit message:

multi-room: add snapweb to snapcast server

This adds the simple web application snapweb to the snapcast server.
The application allows to set properties of snapcast (multiroom) client
such as volume

Change-type: minor
Signed-off-by: emmanuel.jannetti@gmail.com
Connects-to: #319

# This is the commit message #2:

remove use of GIT submodule fri snapweb. Clone it within docker image

# This is the commit message #3:

fix build of snapweb
```

This is how it should look like: 
```
multi-room: add snapweb to snapcast server

This adds the simple web application snapweb to the snapcast server.
The application allows to set properties of snapcast (multiroom) client
such as volume

Change-type: minor
Signed-off-by: emmanuel.jannetti@gmail.com
Connects-to: #319
```

And that's it! Please let me know if this is confusing or not clear. I'm revamping the contributing guidelines and want to make this as clear and easy as possible. Thanks!

**Comment by ejannett** [comment]
> @ejannett looks good, thanks!
> 
> Before merging we'll need you to rebase your branch, what we want is to get all your commits squashed into a single commit and applied on top of the latest changes from our master branch. Reasoning for this is we don't want to clutter the commit history with several small commits for a minor/patch feature and also have multiple `Merge branch 'master' into ...` messages throughout the history.
> 
> Remember the single commit that remains must follow the [commit guidelines](https://sound.balenalabs.io/contributing#commit-guidelines) described in our docs (your first commit already does!).
> 
> The process to do this is not clearly documented at the moment in our docs so I'll leave it here, let me know if you need assistance:
> 
> ```
> git checkout <your-branch>
> git remote add upstream https://github.com/balenalabs/balena-sound.git
> git fetch upstream
> git rebase -i upstream/master
> git push origin <your-branch> --force-with-lease
> ```
> 
> When rebasing, if you made multiple commits then you'll need to `pick` the first one and `squash` the rest. Be sure to clean the commit message and only leave the relevant text.
> 
> This is how it should **not** look like:
> 
> ```
> # This is a combination of 3 commits.
> # This is the 1st commit message:
> 
> multi-room: add snapweb to snapcast server
> 
> This adds the simple web application snapweb to the snapcast server.
> The application allows to set properties of snapcast (multiroom) client
> such as volume
> 
> Change-type: minor
> Signed-off-by: emmanuel.jannetti@gmail.com
> Connects-to: #319
> 
> # This is the commit message #2:
> 
> remove use of GIT submodule fri snapweb. Clone it within docker image
> 
> # This is the commit message #3:
> 
> fix build of snapweb
> ```
> 
> This is how it should look like:
> 
> ```
> multi-room: add snapweb to snapcast server
> 
> This adds the simple web application snapweb to the snapcast server.
> The application allows to set properties of snapcast (multiroom) client
> such as volume
> 
> Change-type: minor
> Signed-off-by: emmanuel.jannetti@gmail.com
> Connects-to: #319
> ```
> 
> And that's it! Please let me know if this is confusing or not clear. I'm revamping the contributing guidelines and want to make this as clear and easy as possible. Thanks!

Hi,

Thank you for the info.
integration requirement make perfect sense.  I still have two tiny questions to be sure we are on the same page :
#1. I do not have to clone a brand new wks , mine should be fine as soon as I have the correct "upstream" remote, right ?  In other words , I can skip the there first steps and jump to the 'rebase' command. My clone looks like this
   ```
% git remote -v             
origin	https://github.com/ejannett/balena-sound.git (fetch)
origin	https://github.com/ejannett/balena-sound.git (push)
upstream	https://github.com/balenalabs/balena-sound.git (fetch)
upstream	https://github.com/balenalabs/balena-sound.git (push)
```
#2. once everything is as expected in my "origin" , shall I wait for a final review or do I trigger the merge to upstream from GitHub directly ?

thank
manu

**Comment by tmigone** [comment]
Hey Manu @ejannett , about point 1 you are correct. 

About point 2,  your `add_snapweb` branch on your `origin` is already linked to this PR so you don't need to do anything else. If you check on the top of this PR it says: 

"merge commits into `ejannett:add_snapweb` from `balenalabs:master`"

So once I approve the PR it will merge the `add_snapweb` branch from your repo (`ejannett/balena-sound`) into the `master` branch on our repo (`balenalabs/balena-sound`). You can then merge the `add_snapweb` branch into your repo's `master` branch if you desire but that's not needed and it won't affect our repo at all. Makes sense?

**Comment by ejannett** [comment]
> Hey Manu @ejannett , about point 1 you are correct.
> 
> About point 2, your `add_snapweb` branch on your `origin` is already linked to this PR so you don't need to do anything else. If you check on the top of this PR it says:
> 
> "merge commits into `ejannett:add_snapweb` from `balenalabs:master`"
> 
> So once I approve the PR it will merge the `add_snapweb` branch from your repo (`ejannett/balena-sound`) into the `master` branch on our repo (`balenalabs/balena-sound`). You can then merge the `add_snapweb` branch into your repo's `master` branch if you desire but that's not needed and it won't affect our repo at all. Makes sense?

Hi,

make perfect sense. thx
manu

**Review by tmigone — APPROVED** [review]
Lgtm! Thanks again for contributing @ejannett!

**Pull request merged**
